src/_macosx.m

```diff
@@ -40,60 +40,84 @@
 static bool keyChangeCapsLock = false;
 /* Keep track of the current mouse up/down state for open/closed cursor hand */
 static bool leftMouseGrabbing = false;
-/* Keep track of whether stdin has been received */
-static bool stdin_received = false;
-static bool stdin_sigint = false;
 // Global variable to store the original SIGINT handler
 static PyOS_sighandler_t originalSigintAction = NULL;
 
-// Signal handler for SIGINT, only sets a flag to exit the run loop
+// Stop the current app's run loop, sending an event to ensure it actually stops
+static void stop_with_event() {
+    [NSApp stop: nil];
+    // Post an event to trigger the actual stopping.
+    [NSApp postEvent: [NSEvent otherEventWithType: NSEventTypeApplicationDefined
+                                         location: NSZeroPoint
+                                    modifierFlags: 0
+                                        timestamp: 0
+                                     windowNumber: 0
+                                          context: nil
+                                          subtype: 0
+                                            data1: 0
+                                            data2: 0]
+             atStart: YES];
+}
+
+// Signal handler for SIGINT, only argument matching for stop_with_event
 static void handleSigint(int signal) {
-    stdin_sigint = true;
+    stop_with_event();
+}
+
+// Helper function to flush all events.
+// This is needed in some instances to ensure e.g. that windows are properly closed.
+// It is used in the input hook as well as wrapped in a version callable from Python.
+static void flushEvents() {
+    while (true) {
+        NSEvent* event = [NSApp nextEventMatchingMask: NSEventMaskAny
+                                   untilDate: [NSDate distantPast]
+                                      inMode: NSDefaultRunLoopMode
+                                     dequeue: YES];
```

Comment:
We've been aligning on `:` it seems like, so suggest doing the same here and below after the `notificationID` indent now.

The name of this function is `flushEvents()` but above you added `stop_with_event()` is there a common form we want to use between them?

```suggestion
        NSEvent* event = [NSApp nextEventMatchingMask: NSEventMaskAny
                                            untilDate: [NSDate distantPast]
                                               inMode: NSDefaultRunLoopMode
                                              dequeue: YES];
```